### ISSUE TITLE
Say all stops every 10th line in editors with NVDA

### ISSUE THREAD

**Issue by Neurrone** [microsoft/vscode#41423]
<ul>
	<li>VSCode Version: Code - Insiders 1.20.0-insider (2bab83a097135cf8c2a1ca1498bb8906f06d2cb3, 2018-01-10T05:13:51.946Z)</li>
	<li>OS Version: Windows_NT x64 10.0.15063</li>
<li>NVDA Version: master-14774,0d94f5dd</li>
	<li><details><summary>Extensions (4)</summary>

Extension|Author (truncated)|Version
---|---|---
vscode-markdownlint|Dav|0.12.1
vscode-rust|kal|0.4.2
rust|rus|0.3.2
markdown-preview-enhanced|shd|0.3.2

</details></li>
</ul>

#### Steps to reproduce:

1. Start NVDA. Create an empty file with the following content:

```markdown
# heading 1

This is a paragraph
of some text
and some more text.

## heading level 2

* item 1
* item 2
* item 3
```

2. Perform a "say all" by pressing insert + down arrow (if NVDA is using a desktop keyboard layout which is the default) or caps lock + a if using laptop keyboard layout when your cursor is anywhere on lines 1-10 (the second-last line "* item 2")

#### Expected behaviour

This should cause NVDA to start reading downwards from the current position to the end of the document.

#### Actual behaviour

NVDA does start reading from the current position, but stops once it reads line 10 when you start reading from lines <= 10.

I remember reading that VS Code's worked with screen readers based on "paging the text". I'm not sure what that means, could this be a side effect?

#### Impact

Reading through long files always stops on every 10th line, which is most noticeable when proofreading long files e.g .md or .tex files.

**Comment by dranov** [comment]
This affects all Monaco editors, not just VSCode.

`IAccessible.Value` always gets set to the 10 line block that includes the focused line (e.g. if line 11 is focused, the value is lines 11-20, or if line 8 is focused, the text is lines 1-10).

![10lines](https://user-images.githubusercontent.com/1622998/43160360-d2245fea-8f7c-11e8-9053-a9502097881c.png)

Bonus: if this issue is fixed (such that `IAccessible.Value` includes the full document text), it would be possible to implement an NVDA app module (app-specific addon) to speak the current line number. (See https://github.com/Microsoft/vscode/issues/52429.)

**Comment by dranov** [comment]
> I remember reading that VS Code's worked with screen readers based on "paging the text". I'm not sure what that means, could this be a side effect?

Background on this: https://github.com/Microsoft/vscode/issues/26730

**Comment by dranov** [comment]
I've (re-)reported the underlying issue to the Chromium bug tracker: https://bugs.chromium.org/p/chromium/issues/detail?id=868830

**Comment by Neurrone** [comment]
Thanks @dranov. Based on the Chromium ticket, it should hopefully be fixed soon.

**Comment by dranov** [comment]
Update from the Chromium ticket:

1. There is a performance issue related to text metrics ('Enables support for querying line breaks and the bounding box of arbitrary character ranges'). I believe this is what @alexandrudima  originally reported and what motivated the introduction of text paging to VSCode. This has not been fixed.

2. Separate from this, there was a recent regression (4 weeks ago) in `<textarea>` scrolling performance when accessibility support is turned on. This regression, which has only affected pre-release versions of Chromium, has been rolled back (and won't affect Chrome 69 stable, due to be released September 4th), but is completely orthogonal to the VSCode issue.

To recap: we've prevented a performance regression from making it into Chrome stable, but the issue which affected VSCode is still there.

I've reported steps to reproduce the underlying issue (which I've also reproduced on Chromium 61.0.3163.0, the version which VSCode uses internally), so hopefully it will get some attention now.

**Comment by dranov** [comment]
Good news! The text metrics performance issue is [now fixed](https://bugs.chromium.org/p/chromium/issues/detail?id=875344) in Chromium. The relevant commit is [here](https://chromium-review.googlesource.com/c/chromium/src/+/1180647). It might take a few months until the fix makes it into Chromium stable and then Electron.

@alexandrudima, is there some way to run VSCode with a particular version of Chromium, so we can make sure this fixed the issue and text paging can now be removed?

**Comment by alexandrudima** [comment]
@dranov There is no easy way to run VSCode, as that requires that there exists an Electron version containing that Chromium version. The closest way to try this would be to run the monaco editor from sources (same source as VS Code)

https://github.com/Microsoft/monaco-editor/blob/master/CONTRIBUTING.md#running-the-editor-from-source

**Comment by alexandrudima** [comment]
@dranov Thank you for testing this! Once we update to a Chromium version >= 70.0.3530.0, we can raise the limits by an order of magnitude.

I should also test if the current limits (10 lines per page, 500 chars before or after the selection) still make sense. Those were put it back when we used an older version, and perhaps we can raise them even on Chromium 61. Would you be willing to look into this and submit a PR to raise the limits?

**Comment by dranov** [comment]
Sure, I'll take a closer look at this and submit a PR 😃

Also, considering the issue here ("reading through long files always stops on *every* 10th line"), besides increasing the current limits, it might be worth it to change the text paging strategy as well. The problem with the current strategy is that, even if you increase the limit to e.g. 100, if you are on line 100 in a larger file and ask NVDA to read until the end, NVDA will just read the current line and stop. Centring the page around the currently selected line might be better. @Neurrone, thoughts?

I'll also try to understand how the Chromium accessibility stuff works -- it seems fundamentally slow, even when text metrics is turned off.

**Comment by alexandrudima** [comment]
@dranov Last time I tried with NVDA on Windows, centering the page around the current selection does not work, as it confuses screen readers terribly. For example, on every arrow down you end up removing the first line of text and adding a new line at the bottom. having an arrow key result in a text change causes terrible confusion to the screen reader which then fails to read out the line.

**Comment by dranov** [comment]
I haven't had the chance to explore this in more depth yet (e.g. understand what's the root cause of the slowness in Chromium or figure out whether it's worth to hack NVDA to understand paging), but I'll dedicate some time for it next week and try to submit a PR.

**Comment by Neurrone** [comment]
CC @jcsteh could you chime in on how NVDA implements paging for editables?

**Comment by isidorn** [comment]
We are currently using Chrome 69.
Assigning to myself to look into this once we do the next electron update and start using chrome 70.

Our progress on electron 6.0 adoption can be found here https://github.com/microsoft/vscode/pull/75802
Once we merge that in we can look into  removing the limits @alexandrudima mentions
@alexandrudima can you maybe provide a code pointer?

**Comment by alexandrudima** [comment]
@isidorn Here is the code that determines what contents should be placed in the textarea. https://github.com/microsoft/vscode/blob/496d06552180275e1431dc3f47e40b4ff8108286/src/vs/editor/browser/controller/textAreaHandler.ts#L203-L233

But we should really test this, because even if Chromium has improved in terms of textarea performance by one magnitude, today they run fine with ~10 lines, maybe they will run fine with ~100 lines... In other words, we should test with super duper large files before removing the limits completely.

**Comment by isidorn** [comment]
@alexandrudima thanks for the pointer. Yes, we can test it thouroughly once we adopt Electron 6.0..
What I find weird is that this issue is still open https://bugs.chromium.org/p/chromium/issues/detail?id=580955
While this one is closed https://bugs.chromium.org/p/chromium/issues/detail?id=875344

**Comment by isidorn** [comment]
We did not yet upgrade to Electron 6.0, thus pushing to next milestone.

**Comment by isidorn** [comment]
We will probably ship elecontr 6 in october, thus I will look into this in November.
As @alexandrudima suggested If this still does not work, we can make this length of the content configurable so that users can easily change this. The downside would be the performance.

**Comment by isidorn** [comment]
I just verified that the latest Chrome that we use still has the issue. And this is expected since https://bugs.chromium.org/p/chromium/issues/detail?id=580955 is still open.

As a workaround we have introduced the following setting
`editor.accessibilityPageSize` as the description of the setting says 
" Controls the number of lines in the editor that can be read out by a screen reader. Warning: this has a performance implication for numbers larger than the default."

So anyone hitting this issue is advised to increase that setting to a larger number (default is 10). For example 1000 works just fine on my machine.

@alexandrudima will measure if we can change the default value, to no longer be 10 but 100 for example.

I am also thinking of having a heuristic to dynamically increase this setting to 100 if we detect a screen reader.

Feedback on the setting name and anything else is very welcome.

**Comment by Neurrone** [comment]
The setting name sounds fine, and I feel that 100 should be a safe default. This should be documented in the accessibility documentation for users with machines that can take a higher value, for example 1000.

**Comment by isidorn** [comment]
Thanks for feedback. I have added it to our Accessibility docs so users are aware of this.
As for changing the default we are open for this, however we will first measure.

**Comment by isidorn** [comment]
Please note that we have changed the `accessibilityPageSize` behavior. If the screen reader is detected we change the default value to be 1000. More details here https://github.com/microsoft/vscode/pull/89173 

This should provide a better out of the box experience. And due to that I plan to remove the tip from our accessibility docs that users need to configure this setting.

**Comment by isidorn** [comment]
This is the actual upstream issue https://bugs.chromium.org/p/chromium/issues/detail?id=1001591&q=10000%20chars&can=2

**Comment by mltony** [comment]
@isidorn, on that chromium issue page someone has commented that it is allegedly fixed now. Would this imply that when new version of chromium is included into vscode, we'd be able to set editor.accessibilityPageSize to infinity?

**Comment by isidorn** [comment]
@mltony in theory yes. In practice no since Chrome still has performance issues with large textArea. Once that get's fixed we will deprecate pageSize setting and always put the whole content of the editor.
@joanmarie might know more about the chrome performance issue

**Comment by joanmarie** [comment]
The performance issue is filed here: https://bugs.chromium.org/p/chromium/issues/detail?id=1076525

In comment 1 put where the problem/crazy-lag/timeout is happening. But we're going to need someone with AXPosition smarts to fix it. (My hopes that this was a simple problem were dashed.)

**Comment by cannona** [comment]
@isidorn Unfortunately, since this change was made, I am seeing a lot of lag when moving around the editor with arrow keys in large files (several hundred lines) while running the latest version of Jaws. We're talking about a good quarter second before Jaws speaks the active character. Setting editor.accessibilityPageSize to 100 resolves the issue.

I was also able to reproduce this issue in NVDA, though it is a shorter lag.

Interestingly, I don't notice the lag when navigating near the top of the file; it seems to get worse the further one moves down the file.

**Comment by isidorn** [comment]
@cannona thanks for letting us know.
Maybe I was too optimistic with setting a default `accessibilityPageSize` to 1000 since there is still the issue https://bugs.chromium.org/p/chromium/issues/detail?id=1076525

Let me reduce that to 500. Hopefully that works out better in practice...

**Comment by isidorn** [comment]
I believe this should work better with the latest changes.
I will go ahead and close this issue. If you still see issues, do let me know and we can reopen

fyi @meganrogge @aiday-mar

**Comment by aiday-mar** [comment]
Hi @isidorn thank you for pinging me on this issue as this is relevant to the native edit context API work too. When I was adopting this API, I was reading through the code for the text area edit context, and I noticed that in this PR https://github.com/microsoft/vscode/commit/4075475d31071d5bae87a50c7e81693346a04c97 we increase the limit of 10 lines per chunk to 500 lines. I am wondering if there is a reason we are increasing to 500 directly within the code, and not changing the default value to 500 directly in the editor options? I am asking as currently the native edit context has not adopted this value of 500 and is still at 10, which could pose problems when the edit context is enabled once again by default (it was disabled due to smoke test issues). If there are no issues with changing directly the default to 500, I would like to do that. Also pinging @meganrogge about this.

**Comment by isidorn** [comment]
I think we can change the default value in the editor. As long as we test it out with NVDA and VoiceOver to make sure we are not breaking any flows (previously having this setting be 500 was slowing down NVDA, so we should make sure there are no regressions)

**Comment by aiday-mar** [comment]
Thanks for the response. It seems to me that the way the current code is written, the accessibility page size is set to 500 on all platforms independent of the OS, so for NVDA users on Windows the default limit is already 500 ever since the commit was made. From this, I think changing the editor option default value should not make any change. I'll wait for Megan's response too.

**Comment by Neurrone** [comment]
I think this issue should remain open, because the solutions are just workarounds for the [underlying bug](https://issues.chromium.org/issues/40688276) in Chrome that causes screen readers to be unresponsive for large documents. It seems like there's been no progress resolving the underlying bug.
